pkg/scheduler/core/equivalence/eqivalence.go

```diff
@@ -70,15 +70,55 @@ func NewCache() *Cache {
 type NodeCache struct {
 	mu    sync.RWMutex
 	cache predicateMap
+	// Generations for predicates, incremented on predicate invalidation.
+	// Created on first update. Use 0 if does not exist.
+	generations map[string]uint64
 }
 
 // newNodeCache returns an empty NodeCache.
 func newNodeCache() *NodeCache {
 	return &NodeCache{
-		cache: make(predicateMap),
+		cache:       make(predicateMap),
+		generations: make(map[string]uint64),
 	}
 }
 
+// NodeCacheSnapshot represents a snapshot of NodeCache reference and generations.
+type NodeCacheSnapshot struct {
+	NodeCache   *NodeCache
+	Generations map[string]uint64
+}
+
+// UpdateNodeCacheSnapshots update node cache snapshots for given nodes. Create node cache if does not exist.
+func (c *Cache) UpdateNodeCacheSnapshots(nodes []*v1.Node, snapshots map[string]*NodeCacheSnapshot) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
```

Comment:
In recent #66862, we've changed to use `sync.Map` for `Cache`, could you please rebase to latest HEAD and use this data structure?

Reply:
Found it's impossible to wrap more than 1 operations on map in a lock, e.g. updating and iterating.

In this snapshotting function, we need to create node cache if not found, then snapshot all node caches. I updated it to use `sync.Map` methods (see below), apparently it's not ok. 

- cannot wrap multiple update operations in one lock
- cannot wrap update operations with iterations in one lock

```
    for _, node := range nodes {
        if _, ok := c.Load(node.Name); !ok {
            c.LoadOrStore(node.Name, newNodeCache())
        }
    }
    c.Range(func(k, v interface{}) bool {
        name := k.(string)
        node := v.(*NodeCache)
        snapshot, ok := snapshots[name]
        if !ok {
            snapshot = &NodeCacheSnapshot{
                Generations: make(map[string]uint64, len(node.generations)),
            }
            snapshots[name] = snapshot
        }
        snapshot.NodeCache = node
        for k, v := range node.generations {
            snapshot.Generations[k] = v
        }
        return true
    })
```

Still using plain Go map with separate locking, what do you think? I think it gives us more freedom.

Reply:
I'm totally OK with either way. The only concern will be performance decrease in machines with 8+ CPUs, let's see how the benchmark says :D